Add initial MPI_T-like profiling support in Mercury

srini009 opened the pull request:
Add MPI-like profiling interface in Mercury. 
Profiling interface is active by default. 
Currently, only one PVAR is exported called "hg_pvar_hg_forward_count" that counts the number of times the HG_Forward call is invoked on this instance.

soumagne commented on src/mercury_prof_interface.h:
```diff
@@ -0,0 +1,37 @@
+#include "mercury_prof_types.h"
+
+/*Initialize and finalize routines */
+HG_PUBLIC hg_return_t HG_Prof_init();
```
Could you please add documentation for every function using the template that is similar to other mercury headers? i.e. something like:
```C
/**
 * <Description>
 *
 * \param param1 [IN/OUT]         <description>
 * \param param2 etc
 *
 * \return HG_SUCCESS or corresponding HG error code
 */
```
Since you added comments in the .c instead, you could just move those comments to the header. We usually don't duplicate function documentation between source and header so having it in one place in the header would be best.

srini009 replied:
@soumagne: I have addressed this issue in the latest commit.

soumagne replied:
Great thanks! Yes that's much easier to review the API.

soumagne commented on src/mercury_prof_interface.h:
```diff
@@ -0,0 +1,126 @@
+ */
+HG_PUBLIC hg_return_t 
+HG_Prof_init();
+
```
That should probably take an hg_class as a parameter. We usually try to avoid using globals.

soumagne replied:
See my previous comment, you should return an `hg_prof_class_t *` here

soumagne commented on src/mercury_prof_interface.h:
```diff
@@ -0,0 +1,126 @@
+ */
+HG_PUBLIC hg_return_t 
+HG_Prof_finalize();
+
```
Same here

soumagne commented on src/mercury_prof_interface.h:
```diff
@@ -0,0 +1,126 @@
+HG_PUBLIC hg_return_t 
+HG_Prof_pvar_handle_alloc(hg_prof_pvar_session_t session, 
+			int pvar_index, void *obj_handle, hg_prof_pvar_handle_t *handle, int *count);
+
```
if you have an alloc routine, you should also have a free routine I expect?

soumagne commented on src/mercury_prof_interface.h:
```diff
@@ -0,0 +1,126 @@
+ */
+HG_PUBLIC int 
+HG_Prof_pvar_get_num();
+
```
Also here

soumagne commented on src/mercury_prof_interface.h:
```diff
@@ -0,0 +1,126 @@
+ */
+HG_PUBLIC hg_return_t 
+HG_Prof_pvar_session_create(hg_prof_pvar_session_t *session);
+
```
Also here

soumagne commented on src/mercury_prof_interface.h:
```diff
@@ -0,0 +1,126 @@
+ */
+HG_PUBLIC hg_return_t 
+HG_Prof_pvar_start(hg_prof_pvar_session_t session, hg_prof_pvar_handle_t handle);
+
```
Same here, I assume if you have a start you will need terminate or end?

soumagne commented on src/mercury.c:
```diff
@@ -22,6 +22,10 @@
+/* PVAR profiling support */
+#include "mercury_prof_pvar_impl.h"
+HG_PROF_PVAR_UINT_COUNTER_DECL_EXTERN(hg_pvar_hg_forward_count);
+
```
As a general comment I think it would be good to see if/how we can avoid declaring global variables.

carns commented:
I would also suggest that the comments for this interface reference a URL (I'm not sure which one off hand) for the MPI_T interface with a brief comment along the lines of "The following interface is patterned off of the MPI_T interface, except that X, Y, Z", where X, Y, an Z are notable philosophical differences (like being oriented around HG classes rather than globals) if there are any.

That way we can lean on the MPI documentation to some degree and also help rationalize the API conventions, since there is some value to reusing terminology for this stuff.

soumagne commented on src/mercury_prof_pvar_impl.h:
```diff
@@ -0,0 +1,89 @@
+/* Increment the value of a PVAR */
+#define HG_PROF_PVAR_COUNTER_INC(name, val) \
+    *(&PVAR_COUNTER_##name) += val;
+
```
@srini009 watching your video, it made me think that you'll need to make sure that all increments are either done using atomics or are protected by a mutex. I think it makes sense to go with atomic types (see https://github.com/mercury-hpc/mercury/blob/master/src/util/mercury_atomic.h) rather than using locks.

carns replied:
Do the performance counters actually need to be atomic?  I had assumed not, but I could be missing something.

soumagne replied:
If you're counting the number of forward calls for example and increment the counter, this has to be protected by a lock or be done atomically.

carns replied:
I definitely see that it could be updated (or accessed) concurrently, just wasn't sure about if that caused danger here or not.  But I guess "+= val" is technically a read/modify/write so you could lose updates if it's not atomic.

I'm on board; we should use the hg atomics.

srini009 commented:
Hey guys @carns, @soumagne: Thank you for your useful comments. I generally agree with all the points that have been put forward. I shall address most if not all comments in the next 2 days. I hope that works!

srini009 commented:
@soumagne, @carns: I have addressed all your comments. Kindly check if this is okay. I apologize for the delay, it took longer than I expected.

carns commented:
> @soumagne, @carns: I have addressed all your comments. Kindly check if this is okay. I apologize for the delay, it took longer than I expected.

Thanks @srini009 !

I just checked two things that I had commented on:

a) did you find a URL or some other reference to put in the comments for the general model the API is using?  It might be there and I'm missing it.

b) there has got to be a better way to implement COUNTER_INC() than looping over an incr() right?  Maybe do a get() and then a cas() instead (with a retry loop if needed)?  Jerome may have another idea.  I'm thinking of the case where the val could be relatively large, like if you are accumulating bytes transmitted or something.  The current code might be slow, and also isn't technically atomic across the macro.

srini009 commented:
Hi @carns,
As for (a), I have addressed the comment by modifying the commit description (see top). Oops, I think you probably meant the code! :) I shall add the same comments in the interface header file as well.

As for (b), indeed, I did give this a thought. As far as atomicity is concerned, hopefully, I'm not mistaken, but I think we don't need to enforce atomicity at the macro level since individual updates are atomic and addition is commutative. I don't think we shall be encountering race conditions here. But you're right, it is incredibly inefficient for large updates related to bytes transferred, etc. Perhaps I can take @soumagne's help here. 

I am excited about this becoming a part of mercury. After that, it is a relatively quick step to introduce a bunch of new PVARs that are insightful.

Regards,

carns commented:
> Hi @carns,
> As for (a), I have addressed the comment by modifying the commit description (see top). Oops, I think you probably meant the code! :) I shall add the same comments in the interface header file as well.
> 
> As for (b), indeed, I did give this a thought. As far as atomicity is concerned, hopefully, I'm not mistaken, but I think we don't need to enforce atomicity at the macro level since individual updates are atomic and addition is commutative. I don't think we shall be encountering race conditions here. But you're right, it is incredibly inefficient for large updates related to bytes transferred, etc. Perhaps I can take @soumagne's help here.
> 
> I am excited about this becoming a part of mercury. After that, it is a relatively quick step to introduce a bunch of new PVARs that are insightful.
> 
> Regards,

Ah!  Yes, that link and explanation in the PR comment is perfect, just drop it in the code.  My thinking is that in the future someone might look at the code and want to change things without realizing that it's intentionally matching some broader conventions.

carns commented:
@soumagne what's the best what to do the performance counter macro?  Do a get(), calculate new value, then cas() (in a loop to retry if the cas fails)?

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+    int hg_prof_is_initialized;     /* Is profiling initialized */
+    int num_pvars;          /* No of PVARs currently supported */
+    hg_prof_pvar_session_t session; /* Is profiling initialized on the session */
+};
```
Maybe I am missing something but you should have a `struct hg_prof_class` here, not `hg_private_class`.

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+hg_prof_set_is_initialized(struct hg_private_class * hg_private_class)
+{
+  hg_private_class->hg_prof_is_initialized = 1;
+}
```
Just a detail but instead of using 1 or 0, there are HG_TRUE and HG_FALSE macros

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+static int 
+hg_prof_get_session_is_initialized(struct hg_prof_pvar_session * session) {
+   return session->is_initialized;
+}
```
Same here, you could return an hg_bool_t instead.

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+  hg_prof_set_is_initialized(hg_private_class);
+  hg_private_class->num_pvars = NUM_PVARS;
+  hg_private_class->session = NULL;
+
```
I don't think you are allowed to do that here :) HG_Prof_init() should return a new hg_prof_class_t * so I would expect you to malloc a new struct hg_prof_class in that routine, fill it and return it.

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+/*---------------------------------------------------------------------------*/
+hg_return_t 
+HG_Prof_finalize(hg_class_t *hg_class) {
+
```
Here instead it should take an hg_prof_class_t *

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+int 
+HG_Prof_pvar_get_num(hg_class_t *hg_class) {
+  struct hg_private_class *hg_private_class = (struct hg_private_class *)(hg_class);
+
```
Same here and following routines

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+
+/*---------------------------------------------------------------------------*/
+hg_return_t 
+HG_Prof_pvar_get_info(hg_class_t *hg_class, int pvar_index, char *name, int *name_len, hg_prof_class_t *var_class, hg_prof_datatype_t *datatype, char *desc, int *desc_len, hg_prof_bind_t *bind, int *continuous) {
```
So that means hg_prof_class_t becomes hg_prof_var_class_t ?

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+
+  struct hg_prof_pvar_session s = *session;
+  unsigned int key = pvar_index;
+  hg_prof_pvar_data_t * val;
```
Please keep declarations at beginning of blocks

soumagne commented on src/mercury_prof_interface.c:
```diff
@@ -0,0 +1,281 @@
+HG_Prof_pvar_handle_free(hg_prof_pvar_session_t session, int pvar_index, hg_prof_pvar_handle_t *handle) {
+
+  if(!hg_prof_get_session_is_initialized(session))
+    return HG_NA_ERROR;
```
that seems like the wrong error code :) HG_NA_ERROR means it's an NA layer error so maybe just HG_INVALID_ARG?

soumagne commented on src/mercury_prof_types.h:
```diff
@@ -0,0 +1,42 @@
+   HG_PVAR_CLASS_SIZE, /* PVAR that represents the size of a given resource at any given point in time */
+   HG_PVAR_CLASS_HIGHWATERMARK, /* PVAR that represents a high watermark value */
+   HG_PVAR_CLASS_LOWWATERMARK /* PVAR that represents a low watermark value */
+} hg_prof_class_t;
```
yes you should rename that one to hg_prof_var_class_t, given the changes that I suggested earlier.

soumagne commented:
@carns @srini009 yes for the performance instead of:
```C
#define HG_PROF_PVAR_COUNTER_INC(name, val)
    addr_##name = (addr_##name == NULL ? hg_prof_get_pvar_addr_from_name(#name): addr_##name);
    for(int i=0; i < val; i++)
        hg_atomic_incr32(addr_##name);
```
you should do something like:
```C
#define HG_PROF_PVAR_COUNTER_INC(name, val) do {
    hg_util_int32_t tmp;

    addr_##name = (addr_##name == NULL ? hg_prof_get_pvar_addr_from_name(#name): addr_##name);
    do {
        tmp = hg_atomic_get32(addr_##name);
    } while (!hg_atomic_cas32(addr_##name, tmp, (tmp + val)));
} while (0)
```
